napari/_vispy/layers/surface.py

```diff
@@ -42,51 +47,74 @@ def __init__(self, layer) -> None:
         self.reset()
         self._on_data_change()
 
+    def _on_texture_change(self):
+        """Update or apply the texture filter"""
+        if self.layer._has_texture and self._texture_filter is None:
+            self._texture_filter = TextureFilter(
+                self.layer.texture,
+                self.layer.texcoords,
+            )
+            self.node.attach(self._texture_filter)
+        elif self.layer._has_texture:
+            self._texture_filter.texture = self.layer.texture
+            self._texture_filter.texcoords = self.layer.texcoords
+
+        if self._texture_filter is not None:
+            self._texture_filter.enabled = self.layer._has_texture
+            self.node.update()
+
     def _on_data_change(self):
-        ndisplay = self.layer._slice_input.ndisplay
-        if len(self.layer._data_view) == 0 or len(self.layer._view_faces) == 0:
-            vertices = None
-            faces = None
-            vertex_values = np.array([0])
-        else:
+        vertices = None
+        faces = None
+        vertex_values = None
+        vertex_colors = None
+        if len(self.layer._data_view) and len(self.layer._view_faces):
             # Offsetting so pixels now centered
             # coerce to float to solve vispy/vispy#2007
             # reverse order to get zyx instead of xyz
             vertices = np.asarray(
                 self.layer._data_view[:, ::-1], dtype=np.float32
             )
-            # due to above xyz>zyx, also reverse order of faces to fix handedness of normals
+            # due to above xyz>zyx, also reverse order of faces to fix
+            # handedness of normals
             faces = self.layer._view_faces[:, ::-1]
-            vertex_values = self.layer._view_vertex_values
 
-        if vertices is not None and ndisplay == 3 and self.layer.ndim == 2:
-            vertices = np.pad(vertices, ((0, 0), (0, 1)))
+            values = self.layer._view_vertex_values
+            if len(values):
+                vertex_values = values
 
-        # manually detach filters when we go to 2D to avoid dimensionality issues
-        # see comments in napari#3475. The filter is set again after set_data!
-        if ndisplay == 2:
-            filt = self.node.shading_filter
-            try:
-                self.node.detach(filt)
-                self.node.shading = None
-                self.node.shading_filter = None
-            except ValueError:
-                # sometimes we try to detach non-attached filters, which causes a ValueError
-                pass
+            colors = self.layer._view_vertex_colors
+            if len(colors):
+                vertex_colors = colors
+
+        # making sure the vertex data is 3D prevents shape errors with
+        # attached filters, instead of trying to attach/detach each time
+        if vertices is not None and vertices.shape[-1] == 2:
+            vertices = np.pad(vertices, ((0, 0), (0, 1)))
```

Comment:
can you add an explicit `mode='constant', constant_values=0` to this call? I think that making it explicit makes the code easier to parse — I'm adding a zero column, not doing any kind of reflection etc. which is (in my experience) the more common use of pad.